finality-aleph/src/metrics.rs

```diff
@@ -127,31 +128,85 @@ pub enum Checkpoint {
     Finalized,
 }
 
-/// TODO(A0-3009): Replace this whole thing.
 #[derive(Clone)]
-pub struct BlockMetrics {
-    inner: Option<Arc<Mutex<Inner>>>,
+pub enum TopBlockMetrics {
+    Prometheus {
+        highest_finalized: Counter<U64>,
+        best: Gauge<U64>,
+    },
+    Noop,
 }
 
-impl BlockMetrics {
-    pub fn noop() -> Self {
-        Self { inner: None }
+impl TopBlockMetrics {
+    pub fn new(registry: Option<&Registry>) -> Result<Self, PrometheusError> {
+        let registry = match registry {
+            None => return Ok(Self::Noop),
+            Some(registry) => registry,
+        };
+        Ok(Self::Prometheus {
+            highest_finalized: register(
+                Counter::new("aleph_highest_finalized_block", "no help")?,
+                registry,
+            )?,
+            best: register(Gauge::new("aleph_best_block", "no help")?, registry)?,
+        })
     }
 
-    pub fn new(registry: &Registry) -> Result<Self, PrometheusError> {
-        let inner = Some(Arc::new(Mutex::new(Inner::new(registry)?)));
+    pub fn update_best(&self, number: BlockNumber) {
+        match self {
+            TopBlockMetrics::Noop => {}
+            TopBlockMetrics::Prometheus { best, .. } => best.set(number as u64),
+        }
+    }
 
-        Ok(Self { inner })
+    pub fn update_highest_finalized(&self, number: BlockNumber) {
+        match self {
+            TopBlockMetrics::Noop => {}
+            TopBlockMetrics::Prometheus {
+                highest_finalized, ..
+            } => {
+                let number = number as u64;
+                if number < highest_finalized.get() {
+                    warn!(target: LOG_TARGET, "Tried to set highest finalized block to a lower number than before. Resetting `highest_finalized` counter.");
+                    highest_finalized.reset();
+                }
+                let delta = number - highest_finalized.get();
+                highest_finalized.inc_by(delta);
+            }
+        }
     }
+}
 
+#[derive(Clone)]
+pub struct BlockMetrics {
+    pub top_block: TopBlockMetrics,
+    timed: Option<Arc<Mutex<TimedBlockMetrics>>>,
```

Comment:
You don't need an option here. add Noop variant to TimedBlockMetrics. You also don't need Arc<Mutex<..>> at all, counters and gauges have some atomics in them. Those changes will help with concurrency issues.

Reply:
I wanted to leave this refactoring to the timed metrics slo task, but I can do this right now. Tho, I have to keep Arc<Mutex<..>> for lru cache in metrics.

Reply:
Also, removing mutex should help with performance when importing many blocks, but if I'm not mistaken the concurrency issue I mentioned could have theoretically still occurred, so I fixed it in other way.